src/metabase/query_processor/middleware/add_implicit_clauses.clj

```diff
@@ -74,10 +84,12 @@
     aggregations :aggregation} :- mbql.s/MBQLQuery]
   ;; if someone is trying to include an explicit `source-query` but isn't specifiying `source-metadata` warn that
   ;; there's nothing we can do to help them
-  (when (and source-query (empty? source-metadata))
-    (when-not qp.i/*disable-qp-logging*
-      (log/warn
-       (trs "Warning: cannot determine fields for an explicit `source-query` unless you also include `source-metadata`."))))
+  (when (and source-query
+             (empty? source-metadata)
+             (qp.store/initialized?))
+    ;; by 'caching' this result, this log message will only be shown once for a given QP run.
+    (qp.store/cached [::should-add-implicit-fields-warning]
+        (log/warn "Warning: cannot determine fields for an explicit `source-query` unless you also include `source-metadata`.")))
```

Comment:
Does it make sense to include a pretty printed version of the query, or something else here to aid with debugging?

Reply:
Good call, I'll add that

Reply:
fixed